EIPS/eip-3369.md

```diff
@@ -0,0 +1,183 @@
+    for i := 0; i < len(mix); i++ {
+        mix[i] = fnvFunc(mix[i], data[i])
+    }
+}
+
+
+// generateDatasetItem combines data from 256 pseudorandomly selected cache nodes,
+// and hashes that to compute a single dataset node.
+func generateDatasetItem(cache []uint32, index uint32, keccak512 hasher) []byte {
+    // Calculate the number of theoretical rows (we use one buffer nonetheless)
+    rows := uint32(len(cache) / hashWords)
+
+    // Initialize the mix
+    mix := make([]byte, hashBytes)
+
+    binary.LittleEndian.PutUint32(mix, cache[(index%rows)*hashWords]^index)
+    for i := 1; i < hashWords; i++ {
+        binary.LittleEndian.PutUint32(mix[i*4:], cache[(index%rows)*hashWords+uint32(i)])
+    }
+    keccak512(mix, mix)
+
+    // Convert the mix to uint32s to avoid constant bit shifting
+    intMix := make([]uint32, hashWords)
+    for i := 0; i < len(intMix); i++ {
+        intMix[i] = binary.LittleEndian.Uint32(mix[i*4:])
+    }
+    // fnv it with a lot of random cache nodes based on index
+    for i := uint32(0); i < datasetParents; i++ {
+        parent := fnv0(index^i, intMix[i%16]) % rows
+        fnvHash(intMix, cache[parent*hashWords:], fnv0)
+    }
+    // Flatten the uint32 mix into a binary one and return
+    for i, val := range intMix {
+        binary.LittleEndian.PutUint32(mix[i*4:], val)
+    }
+    keccak512(mix, mix)
+    return mix
+}
+
+// hashimoto aggregates data from the full dataset in order to produce our final
+// value for a particular header hash and nonce.
+func hashimoto(hash []byte, nonce uint64, size uint64, lookup func(index uint32) []uint32) ([]byte, []byte) {
+    // Calculate the number of theoretical rows (we use one buffer nonetheless)
+    rows := uint32(size / mixBytes)
+
+    // Combine header+nonce into a 64 byte seed
+    seed := make([]byte, 40)
+    copy(seed, hash)
+    binary.LittleEndian.PutUint64(seed[32:], nonce)
+
+    seed = crypto.Keccak512(seed)
+    seedHead := binary.LittleEndian.Uint32(seed)
+
+    // Start the mix with replicated seed
+    mix := make([]uint32, mixBytes/4)
+    for i := 0; i < len(mix); i++ {
+        mix[i] = binary.LittleEndian.Uint32(seed[i%16*4:])
+    }
+    // Mix in random dataset nodes
+    temp := make([]uint32, len(mix))
+
+    for i := 0; i < loopAccesses; i++ {
+        parent := fnvA(uint32(i)^seedHead, mix[i%len(mix)]) % rows
+        for j := uint32(0); j < mixBytes/hashBytes; j++ {
+            copy(temp[j*hashWords:], lookup(2*parent+j))
+        }
+        fnvHash(mix, temp, fnvB)
+    }
+    // Compress mix
+    for i := 0; i < len(mix); i += 4 {
+        mix[i/4] = fnvC(fnvD(fnvE(mix[i], mix[i+1]), mix[i+2]), mix[i+3])
+    }
+    mix = mix[:len(mix)/4]
+
+    digest := make([]byte, common.HashLength)
+    for i, val := range mix {
+        binary.LittleEndian.PutUint32(digest[i*4:], val)
+    }
+    return digest, crypto.Keccak256(append(seed, digest...))
+}
+```
+
+
```

Comment:
```suggestion
```
Everything in an EIP must be CC0 licensed.  This section is optional, so if you cannot get a CC0 licensed version of this code then just remove the section.